multiuser/permission/che-multiuser-permission-workspace/src/main/java/org/eclipse/che/multiuser/permission/workspace/server/model/impl/WorkerImpl.java

```diff
@@ -54,7 +55,8 @@
           "SELECT worker "
               + "FROM Worker worker "
               + "WHERE worker.userId = :userId "
-              + "AND worker.workspaceId = :workspaceId ")
+              + "AND worker.workspaceId = :workspaceId ",
+      hints = {@QueryHint(name = "eclipselink.query-results-cache", value = "true")})
```

Comment:
Does not it lead to a case where a user is able to request JSON RPC methods even after permissions removing?

Reply:
Good question.
I'll check

Reply:
Eclipse Link docs explicitly mention the possibility of getting stale results with the query results cache. I'd imagine this would be more prevalent in clustered instance of che-server.

Reply:
@sleshchenko @metlos I think we safe here. What I've checked:
- UX. I've shared the workspace with another user. Then I've checked that another user can see it. After that, I've removed the second user from the workspace and as a result, another user lost his access to the workspace
- I've checked how EclipseLink is work during WorkerImpl delete operation https://github.com/eclipse/che/pull/13034/files#diff-8590cbe826cfc6e7a8269573c5922b86R200 
It looks like it's invalidating all caches related to WorkerImpl, because of next calls are going directly to the database https://github.com/eclipse/che/pull/13034/files#diff-8590cbe826cfc6e7a8269573c5922b86R201-R202

As a safe trigger, I can add time-based cache invalidation.  Let's say 10 minutes.
WDYT?